src/main/kotlin/org/jetbrains/research/ml/coding/assistant/tree/Numbering.kt

```diff
@@ -0,0 +1,58 @@
+package org.jetbrains.research.ml.coding.assistant.tree
+
+import com.github.gumtreediff.tree.ITree
+import com.github.gumtreediff.tree.TreeContext
+import com.github.gumtreediff.tree.TreeUtils
+import com.intellij.openapi.application.ApplicationManager
+import com.intellij.openapi.util.Key
+import com.intellij.psi.PsiElement
+import org.jetbrains.research.ml.coding.assistant.util.postOrder
+import org.jetbrains.research.ml.coding.assistant.util.preOrder
+
+abstract class Numbering {
+
+    // To be consistent with GumTree [TreeUtils]
+    class PsiTreeUtils {
+        companion object {
+            private val psiId: Key<Int> = Key("ID")
+
+            val PsiElement.id: Int?
+                get() = ApplicationManager.getApplication().runReadAction<Int?> {
+                    this.getCopyableUserData(psiId)
+                }
+
+            fun numbering(iterable: Iterable<PsiElement>) {
+                iterable.forEachIndexed { i, t ->
+                    t.putCopyableUserData(psiId, i)
+                }
+            }
+        }
+    }
+
+    protected abstract fun PsiElement.iterable(): Iterable<PsiElement>
+    protected abstract fun ITree.iterable(): Iterable<ITree>
+
+    fun number(psiElement: PsiElement) {
```

Comment:
Should we make it `private`?

Reply:
I thought we want to have a possibility to number only one tree